Graph Javascript SDK Example (#779)
Macrometacorp/docs: docs/graphs/graph-examples/graph-sdk-example.md

---
title: Graph SDK Example
sidebar_position: 30
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

This example shows the ability of graph databases in analyzing and offering recommendations for items that are commonly purchased together. By establishing connections between customer and item vertices using edge collections, the graph is able to accurately deliver tailored recommendations to customers who have bought a particular item.

![Create a collection](/img/graphs/sdk-graph-image.png)

> In the graph that we created ,customers Dino and Durga bought the same items. The graph is able to recommend items to Dallas based on the items that Dino and Durga bought. This is a simple example of how graph databases can be used to offer recommendations to customers.

Reviewer: ```suggestion
In the graph that we created, customers Dino and Durga bought the same items. The graph recommends items to Dallas based on the items that Dino and Durga bought. This is a simple example of how graph databases can offer recommendations to customers.
```

- Install the Macrometa [javascript SDK](../../sdks/install-sdks.md).

:::note
To check your Node.js version, run the following command in your terminal:

```bash
node -v
```
:::

### Create a Graph

<Tabs groupId="operating-systems">
<TabItem value="js" label="Javascript">

```js
const jsc8 = require("jsc8");

client = new jsc8({
  url: "https://play.paas.macrometa.io",
  apiKey: "xxxxx",
  fabricName: "_system",
});

// Edge collections and vertices must be created before running this script.
async function createGraph() {
  const response = await client.createGraph("grocery-graph", {
    edgeDefinitions: [
      {
        // Edge collection name --> This collection holds relationships between vertices
        collection: "groceryOrders",
        // Vertex collection that is used as the start vertex of the edge
        from: ["groceryCustomers"],
        // Vertex collection that is used as the start vertex of the edge
        to: ["groceryItems"],
      },
    ],
  });
  console.log(response);
}
createGraph();
```

</TabItem>

</Tabs>

After we have created the graph, we can start querying it.

> ##  How to Query the Graph

Reviewer: ```suggestion
##  Query the Graph
```

> ### Query Explanation

Reviewer: ```suggestion
```

Query for the recommendations for a customer can be done using the following query:

<Tabs groupId="operating-systems">
<TabItem value="js" label="C8QL">

```js

// Query: Recommend the most popular item in the grocery store based on the number of users who purchased it
LET doc = (FOR d in groceryItems FILTER d.name == @name RETURN d)[0]
FOR user IN 1..1 INBOUND doc groceryOrders
    FOR item IN 1..1 OUTBOUND user groceryOrders
        COLLECT i=item.name WITH COUNT INTO c
        FILTER i != @name
        SORT c DESC
        // LIMIT 2
        RETURN {"item": i, "count": c}   // item -> number of users who purchased it
                                         // count -> number of times it was purchased
```

</TabItem>

</Tabs>

The `@name` is a bind variable that can be set to any item name. In this example we can set it to `peanut butter`. First we are filtering by item name and then collecting all the users who purchased that item. Then we are collecting all the items that were purchased by those users. We are filtering out the item that we are looking (`peanut butter`) for and sorting the result by the number of times it was purchased.

### Create Query Worker

> It is possible to create a Query Worker on Macrometa GDN using console, javascript SDK, CLI or using the REST API.

Reviewer: ```suggestion
You can create a query worker on Macrometa GDN using console, the JavaScript SDK, CLI, or using the REST API.
```
